packages/webapp/src/containers/Finances/useTransactions.js

```diff
@@ -0,0 +1,241 @@
+  const groupedByRevenueTypeSales = groupBy(
+    filteredSales,
+    ({ revenue_type_id }) => revenue_type_id,
+  );
+  const groupedTransactions = [];
+
+  Object.keys(groupedByRevenueTypeSales).forEach((revenueTypeId) => {
+    const groupedByDateSales = groupBy(
+      groupedByRevenueTypeSales[revenueTypeId],
+      ({ sale_date }) => sale_date,
+    );
+    const revenueType = revenueTypes.find(
+      (revenueType) => revenueType.revenue_type_id == revenueTypeId,
+    );
+
+    Object.keys(groupedByDateSales).forEach((date) => {
+      const revenueItems = mapSalesToRevenueItems(
+        groupedByDateSales[date],
+        revenueTypes,
+        cropVarieties,
+      );
+      groupedTransactions.push({
+        icon: revenueType?.farm_id ? 'OTHER' : revenueType?.revenue_translation_key,
+        date,
+        transactionType: transactionTypeEnum.revenue,
+        typeLabel: revenueType?.farm_id
+          ? revenueType?.revenue_name
+          : i18n.t(`revenue:${revenueType?.revenue_translation_key}`),
+        amount: revenueItems.reduce((sum, item) => sum + item.totalAmount, 0),
+        items: revenueItems,
+      });
+    });
+  });
+
+  return groupedTransactions;
+};
+
+export const buildTransactions = ({
+  sales,
+  tasks,
+  expenses,
+  expenseTypes,
+  revenueTypes,
+  taskTypes,
+  cropVarieties,
+  users,
+  dateFilter,
+  expenseTypeFilter,
+  revenueTypeFilter,
+}) => {
+  const transactions = [
+    ...buildLabourTransactionsFromTasks({ tasks, taskTypes, users, dateFilter, expenseTypeFilter }),
+    ...buildExpenseTransactions({ expenses, expenseTypes, dateFilter, expenseTypeFilter }),
+    ...buildRevenueTransactions({
+      sales,
+      revenueTypes,
+      cropVarieties,
+      dateFilter,
+      revenueTypeFilter,
+    }),
+  ];
+
+  const sortedTransactions = transactions.sort((a, b) => {
+    if (a.date > b.date) {
+      return -1;
+    }
+    if (a.date < b.date) {
+      return 1;
+    }
+    return 0;
+  });
+
+  return sortedTransactions;
+};
+
+const useTransactions = ({ dateFilter, expenseTypeFilter, revenueTypeFilter }) => {
+  const sales = useSelector(salesSelector);
+  const tasks = useSelector(tasksSelector);
+  const expenses = useSelector(expenseSelector);
+  const expenseTypes = useSelector(allExpenseTypeSelector);
+  const revenueTypes = useSelector(allRevenueTypesSelector);
+  const taskTypes = useSelector(taskTypesSelector);
+  const cropVarieties = useSelector(cropVarietiesSelector);
+  const users = useSelector(userFarmsByFarmSelector);
+
+  const transactions = useMemo(
+    () =>
+      buildTransactions({
+        sales,
+        tasks,
+        expenses,
+        expenseTypes,
+        revenueTypes,
+        taskTypes,
+        cropVarieties,
+        users,
+        dateFilter,
+        expenseTypeFilter,
+        revenueTypeFilter,
```

Comment:
I wonder if it would make sense to return all transactions (rather than return no transactions) if `revenueTypeFilter` and `expenseTypeFIlter` are not provided. I noticed this is how `dateFilter` behaves, and it will also save type selectors having to get imported in the transaction list container to populate the default (unfiltered) view when they are already also called here.

Maybe this isn't the best way to write it, but the behaviour could be equivalent to something like this:

```suggestion
        expenseTypeFilter: expenseTypeFilter || [
          ...expenseTypes.map(({expense_type_id}) => expense_type_id),
          'LABOUR_EXPENSE',
        ],
        revenueTypeFilter: revenueTypeFilter || revenueTypes.map(({revenue_type_id}) => revenue_type_id),
```

Reply:
I had originally implemented it this way, but then I figured we'd need a way for the user to enable all expense type filters and no revenue type filters to only view expenses and no revenues, or the other way round, and in that scenario we'd need some way for the UI to explicitly say we shouldn't show any revenue types or any expense types (which would be the empty arrays with the current implementation).

Reply:
Empty arrays are truthy so they should still work with the above (they won't default to the righthand side)

Reply:
You're absolutely right! I've just updated this 🙂